### ISSUE TITLE
Multiplying tall matrices is extremely slow

### ISSUE THREAD

**Issue by florianwechsung** [mathnet/mathnet-numerics#424]
I have noticed that multiplying a tall matrix with a small matrix is extremely slow and does not scale linearly in the tallness of the matrix. 

Example:
Let n>>d, T\in R^{n\times d} and S\in R^{d\times d}, then the time to calculate the product TS should scale linearly in n.

Code: 

        [Test]
        public void MatrixMultiplyTest()
        {
            var ns = new List<int> {5000, 10000, 20000, 40000};
            var d = 10;
            var small = DenseMatrix.CreateRandom(d, d, new ContinuousUniform());
            foreach (var n in ns)
            {
                var tall = DenseMatrix.CreateRandom(n, d, new ContinuousUniform());
                var sw1 = new System.Diagnostics.Stopwatch();
                sw1.Start();
                var prod1 = tall.Multiply(small);
                sw1.Stop();
                var sw2 = new System.Diagnostics.Stopwatch();
                sw2.Start();
                var prod2 = DenseMatrix.OfColumnVectors(tall.EnumerateRows().Select(row => small.Transpose().Multiply(row))).Transpose();
                sw2.Stop();
                Console.WriteLine("n = " + n + ", MathNet: " + sw1.ElapsedMilliseconds/1000.0 + "s, Naive: " + sw2.ElapsedMilliseconds/1000.0 + "s.") ;
                Assert.That(prod1.ToArray(), Is.EquivalentTo(prod2.ToArray()));
            }
        }

Output: 

        n = 5000, MathNet: 0,208s, Naive: 0,026s.
        n = 10000, MathNet: 1,13s, Naive: 0,01s.
        n = 20000, MathNet: 7,714s, Naive: 0,021s.
        n = 40000, MathNet: 66,842s, Naive: 0,064s.

I am especially confused how enumerating over the rows can be so much faster than the normal matrix multiplication and why the runtime does not scale linearly.

**Comment by Aixile** [comment]
Math.Net use divide and conquer to do matrix multiplication by default.

In function CacheObliviousMatrixMultiply, the structure is like

```
static void CacheObliviousMatrixMultiply(Transpose transposeA, Transpose transposeB, float alpha, float[] matrixA, int shiftArow, int shiftAcol, float[] matrixB, int shiftBrow, int shiftBcol, float[] result, int shiftCrow, int shiftCcol, int m, int n, int k, int constM, int constN, int constK, bool first)
{
	if (m + n <= Control.ParallelizeOrder){
		//do submatrix multiplication
	}else{
		int n2=n/2;int m2=m/2;int k2=k/2;
		//do recusion
	}
}
```

If we multiply a m\*k matrix and a k\*n matrix (m>>n,k), n,k will reach to zero soon, but it will continue to do recursion, since m is bigger than 64. 
Obviously, when either n or k is zero, the answer will not be changed.
For instance, when computing the product of a 40000\*10 matrix and a 10\*10 matrix, the submatrix multiplication part is called 917584494 times, but only 3998543 of them actually update the final answer.

After fixing this issue, 
```
n = 5000, MathNet: 0.02s, Naive: 0.022s.
n = 10000, MathNet: 0.007s, Naive: 0.024s.
n = 20000, MathNet: 0.013s, Naive: 0.051s.
n = 40000, MathNet: 0.027s, Naive: 0.113s.
```

**Comment by cdrnet** [comment]
Thanks a lot for the analysis!

**Comment by cdrnet** [comment]
This is released in v3.13.

@florianwechsung can you confirm this has improved your use case as well?

**Comment by florianwechsung** [comment]
Yes, works perfectly! Thanks @Aixile
